Content.Server/Paint/PaintSystem.cs

```diff
@@ -0,0 +1,97 @@
+using Content.Shared.Popups;
+using Content.Shared.Paint;
+using Content.Shared.Interaction;
+using Content.Server.Chemistry.Containers.EntitySystems;
+using Robust.Shared.Audio.Systems;
+using Content.Shared.Humanoid;
+using Content.Shared.SubFloor;
+
+namespace Content.Server.Paint;
+
+/// <summary>
+/// Colors target and consumes reagent on each color success.
+/// </summary>
+public sealed class PaintSystem : SharedPaintSystem
+{
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SolutionContainerSystem _solutionContainer = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<PaintComponent, AfterInteractEvent>(OnInteract);
+    }
+
+    // When paint is used on an entity it will apply the color to the entity provided TryPaint returns true.
+    private void OnInteract(Entity<PaintComponent> entity, ref AfterInteractEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!args.CanReach || args.Target is not { Valid: true } target)
+            return;
+
+        if (!TryComp(args.Used, out PaintComponent? component))
+            return;
+
+        if (TryPaint(entity, target, args.User, args.Used))
+        {
+
+            if (HasComp<AppearanceComponent>(target))
+                RemComp<AppearanceComponent>(target);
+
+            EnsureComp<PaintedComponent>(target);
+            AddComp<AppearanceComponent>(target);
+
+            if (!TryComp(target, out PaintedComponent? paint))
+                return;
+            else
+            {
+                paint.Color = component.Color; // set the target color to the color specified in the spray paint yml.
+                _audio.PlayPvs(component.Spray, args.Used);
+                paint.Enabled = true;
+                UpdateAppearance(target, paint);
+                Dirty(target, paint);
+                args.Handled = true;
+                return;
+            }
+        }
+
```

Comment:
This looks like it's going to be very destructive to visualizers if you're bulldozing appearancecomponent so I really don't think you should be doing that. It won't reconstruct stuff like randomspritevisualizers + will break photos.

Also you don't need to has Comp for RemComp it returns a bool whether the component is removed.

The paint's supposed to be cleanable.

Reply:
i am no longer bulldozing appearance. Instead i am ensuring appearance and using appearance.SetData to make sure the visualizer works regardless of the current appearance of the entity.